### PR TITLE
Feat: added config code example

### PR DESCRIPTION
**Pull request by MattBuckley-TfN**
# Changes

Added code example showing how to create configuration classes with `BaseConfig`.

# Tasks

- [x] Have unittests been added (testing individual functions and their edge cases) - NA
- [x] Has documentation (including user guide) been updated - Yes
- [x] Have new dependencies been added (or changed) in relevant `requirements.txt` - No
- [ ] Code linting, tests and other workflows pass

----
📚 Documentation preview 📚: https://caftoolkit--168.org.readthedocs.build/en/168/

### CODE REVIEW

**Comment by MattBuckley-TfN** [comment]
TODO: exclude examples folder from being ran as tests, this should be ran by the docs building process

File: examples/run_config.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+# Third Party
+import pydantic
+from pydantic import dataclasses
+
+# Local Imports
+import caf.toolkit as ctk
+
+folder = pathlib.Path().parent
+
```
Comment (BenTaylor-TfN) [inline examples/run_config.py:29]: The way this renders it looks like this code is part of the "See Also" section. I think this is just setup to make the rest of the code work? 

Should we add a heading to make it clear that this is just setup for the rest of the code?

File: examples/run_config.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
+# -*- coding: utf-8 -*-
+"""
+Using :class:`~caf.toolkit.BaseConfig`
+======================================
+
+The :class:`~caf.toolkit.BaseConfig` class in caf.toolkit is designed to load
+and validated YAML configuration files. This example shows how
```
Comment (BenTaylor-TfN) [inline examples/run_config.py:7]:
```suggestion
and validate YAML configuration files. This example shows how
```

File: examples/run_config.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+to create child classes of BaseConfig to load parameters.
+
+See Also
+--------
+:class:`~caf.toolkit.arguments.ModelArguments`
+    for producing command-line arguments from a BaseConfig class.
+"""
```
Comment (BenTaylor-TfN) [inline examples/run_config.py:14]:
```suggestion
See Also
--------
You may also want to check out the :class:`~caf.toolkit.arguments.ModelArguments` class for producing command-line arguments from a BaseConfig class.
"""
```

File: examples/run_config.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+# Local Imports
+import caf.toolkit as ctk
+
+folder = pathlib.Path().parent
+
+# %%
+# Basic
+# -----
+#
+# This example shows how to create a simple configuration file, without nesting, and
```
Comment (BenTaylor-TfN) [inline examples/run_config.py:34]: Trying to keep the language a bit simpler for the basic example
```suggestion
# This example shows how to create a simple configuration file
```

File: examples/run_config.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+# with different types of parameters. The majority of Python built-in types can be used,
+# additionally dataclasses and many "simple"[#simple]_ custom types can be used.
+#
+# .. seealso ::
+#       :ref:`Extra Validation` for information on more complex validation.
+#
+# The :class:`pydantic.DirectoryPath` and :class:`pydantic.FilePath` types both return
+# :class:`pathlib.Path` objects after validating that the directory, or file, exists.
```
Comment (BenTaylor-TfN) [inline examples/run_config.py:42]: I think this would be better is a "note" box, similar to the "see also" above. Comment below has more detail

File: examples/run_config.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+#
+# .. literalinclude:: /../../examples/basic_config.yml
+#       :language: yaml
+#       :caption: Config file: examples/basic_config.yml
+#
+# Below shows how to load the config file and displays the class as text.
+
+parameters = Config.load_yaml(folder / "basic_config.yml")
+print(parameters)
+
```
Comment (BenTaylor-TfN) [inline examples/run_config.py:65]: I think it would be good at this point to include the "Note" box, which basically notes a few interesting things about the example above, e.g.,
* How the `DirectoryPath` and `FilePath` types work
* That the names in the Config class need to be the same in the YAML, but the order can be different.

I think it's better after the example so that reader doesn't get overwhelmed with too much information when reading this example.

File: examples/run_config.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    regex_text: str = pydantic.Field(pattern=r"[a-z]+")
+
+    # Iterable restrictions e.g. lists and tuples
+    short_list: list[int] = pydantic.Field(min_length=2, max_length=5)
+
+
+# %%
+# For more complex validation pydantic allow's custom methods to be defined
+# to validate individual fields (:func:`pydantic.field_validator`), or the
+# class as a whole (:func:`pydantic.model_validator`).[#valid]_
```
Comment (BenTaylor-TfN) [inline examples/run_config.py:160]: Is the `[#valid]_` text at the end a typo? It shows up in the render and I don't know why

**Review by BenTaylor-TfN — COMMENTED** [review]
Noting that the index file has 3 examples, but 2 of them are examples of examples. If we want people to use this page now, the first two might be confusing! 
https://caftoolkit--168.org.readthedocs.build/en/168/_generated/examples/index.html

File: examples/run_config.py (hunk 7 above)
Comment (MattBuckley-TfN) [reply examples/run_config.py:160]: It's supposed to link to the footnote, I must have done something wrong with the syntax though. I'll fix it

**Comment by MattBuckley-TfN** [comment]
Addressed Ben's comments so merging
